### PR TITLE
- Discover the root device

### PR DESCRIPTION
**Pull request by rjschwei**
+ With asynchronous device detection we can no longer assume that
    the root is on /dev/sda, the previous default device.

### CODE REVIEW

**Review by jgleissner — APPROVED** [review]
LGTM. I suspect this won't work with DM volumes though.

**Review by bear454 — APPROVED** [review]

File: lib/AzureMetadata/AzureDisk.pm [hunk 1]
```diff
@@ -17,20 +17,55 @@ package AzureDisk;
...
+use File::Basename;
 use POSIX;
 
+
 our @ISA    = qw (Exporter);
 our @EXPORT_OK = qw (
     getVHDDiskTag
 );
 
+sub get_root_device {
```
Comment (schaefi) [inline lib/AzureMetadata/AzureDisk.pm:29]: looks pretty complex to me. Any reason why:

```
findmnt -n -f -o SOURCE /
```

does not work for you ?

Comment (schaefi) [reply lib/AzureMetadata/AzureDisk.pm:29]: Other than that I would care for:


Subroutine "get_root_device" does not end with "return" at line 29, column 1.  See page 197 of PBP.  (Severity: 4)

Return value of "open" ignored at line 40, column 13.  Check the return value of "open" for success.  (Severity: 3)

Comment (rjschwei) [reply lib/AzureMetadata/AzureDisk.pm:29]: Was trying to avoid shelling out.

**Comment by ikapelyukhin** [comment]
@rjschwei I guess I should port this code to the new python implementation?

File: lib/AzureMetadata/AzureDisk.pm (hunk 1 above)
Comment (rjschwei) [reply lib/AzureMetadata/AzureDisk.pm:29]: @ikapelyukhin yes, the python implementation needs to find the root device, again preferably without shelling out

Comment (schaefi) [reply lib/AzureMetadata/AzureDisk.pm:29]: what's wrong on shelling out ? I think the code here could result in some issues that will also be hard to debug

a) $block_device_name = basename($block_device)

    You assume the name of the block device symlink in /sys/block is also used in the name of the
    partitions on that block device. This is an assumption that is not necessarily correct, especially
    on devices that are mapping, like dm, lvm and friends

b) Next you raw read any partition device to get the major and minor and compare that with the
    stat result from /. Why not again call stat on the device to get that information ? It looks relatively complex to me.

I mean the entire block of 28lines of code with no test could be replaced by one findmnt. That should be worth a thought, shouldn't it ?

Thanks

Comment (schaefi) [reply lib/AzureMetadata/AzureDisk.pm:29]: Really don't understand why even the python code should not shell out. findmnt belongs to util-linux, installed everywhere and required for any OS to function (provides e.g mount). Please can we stop re-inventing the wheel
Doesn't make any sense to me

**Review by schaefi — CHANGES_REQUESTED** [review]
I'm not in favor of this root device detection for the given reasons

**Comment by schaefi** [comment]
> @rjschwei I guess I should port this code to the new python implementation?

I would consider it a big mistake

**Comment by ikapelyukhin** [comment]
Well, I don't really have an opinion. I understand both the desire to not reinvent and not to have any dependencies :slightly_smiling_face:

**Comment by schaefi** [comment]
imho there is no dependency issue at all. As I mentioned findmnt belongs to util-linux which apart from other core tools provides the mount utility. Even without a requires tag in the spec file of azuremetada you can be sure this tool is present on linux. If it's not you have more problems than just a missing findmnt ;)

**Comment by schaefi** [comment]
rpm -ql util-linux | grep -E "mnt|mount"

```
/bin/findmnt
/bin/mount
/bin/umount
/usr/bin/findmnt
/usr/bin/mount
/usr/bin/mountpoint
/usr/bin/umount
...many more
```

**Comment by rjschwei** [comment]
OK, since the conversation was split I just merged the other (Python) PR as there was no no indication it was waiting for this conversation.

Yes, the implementation not using findmnt has issues with mapping representations, which we do not use in our images but a customer building their own images could. At that point the question arises how valuable disktag query is, but never the less it should at least read the proper device even if it just returns jibberish.

Long story short let's use `findmnt`

@ikapelyukhin would you please update the Python code, I think then we are ready with the new implementation to move towards replacing the Perl implementation. SInce the Python code is checked in I am going to close this PR.

**Comment by ikapelyukhin** [comment]
@rjschwei OK!

**Comment by schaefi** [comment]
A good decision, thanks guys :+1: If there is anything I can help with with e.g findmnt just let me know. Depending on how complex the root device of the machine is used it might be needed to have further processing on the result of findmnt. This is the case of e.g root is a multipath device. But in any case the first step is to get a device from findmnt, anything afterwards is handling of special mappings based on that device.

From my experience with all this, this was the safest way to handle 95% of the cases  and also the rest 5% corner cases in a way that the code stays readable and testable.

Again thanks
